move Parmatch.Pattern_head into Patterns.Head for a new Patterns module

gasche opened the pull request:
The aim is to also move the Simple/Half_simple/General stuff from
matching, but we need to split in those modules the part that are
purely structural (they go in Patterns) and the parts that are
actually compilation logic (Half_simple.of_clause), those stay in
Matching.

gasche committed:
matching: pass explicit arity to Default_environment.specialize

gasche committed:
[REVIEW REQUIRED] matching: remove the OrPat exception by handling Tpat_or on the claler side

gasche committed:
matching: matcher_* take Simple.pattern arguments

gasche committed:
matching: fix a very old bug in Default_environment.specialize_matrix

The bug was inserted in 842eb76ef7f5f3f85ce4fac22acf201fb237a238 in
2002. It breaks specialization of default matrices, so in theory it
should not lead to wrong-code production, but because it can create
default matrices of distinct arity in the same environment it should
be able to observe assertion failures in code hitting the bug.

I found the bug while upgrading the code to use better types. The
input of specialize_matrix has non_empty rows, but the output may be
empty -- and this code becomes ill-typed when input and output have
distinct types. Yay typing.

gasche committed:
matching: specialize_matrix uses non-empty rows

gasche committed:
matching: consolidate all matcher_ functions in a single matcher_head

The `matcher_of_pattern` function is a temporary measure to reduce the
invasiveness of the patch, and make it easier to review.

(Note for reviewers: in the previous version the Record case had
a funny handling of Any, but it is in fact equivalent to just adding
omegas as we now do in all cases.)

There are two obvious directions for improvement:

- Get rid of matcher_of_pattern and pass a head directly to the
  various make_matching_* functions.

- Try to factorize this code with ctx_matcher which, it is now
  obvious, does essentially the same thing.

Another, less immediate area of attack would be to consider
a presentation of Pattern_head.t where the Any case can be statically
ruled out -- maybe the description could have two levels, one
isomorphic to option (Any or not?) and one for non-any heads.

gasche committed:
matching: (matcher Any p rem = rem) does the right thing

gasche committed:
matching: [minor] Default_environment.specialize takes a head directly

gasche committed:
matching: refine the types in Context.ctx_matcher

gasche committed:
matching: finally, merge matcher and Context.ctx_matcher

gasche committed:
matching: use heads in the make_*_matching specialization calls

gasche committed:
matching: factorize the make_*_matching functions

Note: this commit makes a minor improvement to the location used to
force (lazy ..) arguments.

gasche committed:
move Parmatch.Pattern_head into Patterns.Head for a new Patterns module

The aim is to also move the Simple/Half_simple/General stuff from
matching, but we need to split in those modules the part that are
purely structural (they go in Patterns) and the parts that are
actually compilation logic (Half_simple.of_clause), those stay in
Matching.

gasche committed:
[minor] rename Non_empty_clause.{map_head => map_first}

trefis commented:
IIUC this currently only moves `Parmatch.Pattern_head` to `Patterns.Head` however:
- you did not `git add wherever/patterns.ml`
- you're reverting part the changes you pushed to #15 (you're reverting the `combine_handler` changes)

trefis commented:
> I think the present PR should be considered in tandem with the rematch-more-types branch (on top of #15) -- if we agree to have general pattern-concepts in a dedicated module, I think it could/should be part of the upstream PR for more-types.

I agree that the file is good to have.
Some things to keep in mind though:
- do we really want to include this in #15 or should we send it separately? What's going to be the easier to review?
- I've been planning for a while to move `Typecore.type_pat` to its own `Typepat` file (or something similar). Would it make sense to move it to `Patterns.type_expect`? (Probably not as this would introduce a dependency loop since typing depends on parmatch and parmatch depends on the Heads module)

gasche commented:
My gut feeling is that once this PR stabilizes, the total-diff will be more readable than the current total-diff of #15: a lot of new code will be in its own new module (rather than spliced within Matching), and the changes to Matching itself will look smaller / more local.

gasche commented:
(No, I don't think that we should put typing-patterns in that file, I think a separate `typepat.ml` would be the right choice. The idea of this file is to capture the *data space* related to patterns (and subcategories of patterns).)

trefis commented:
Note: my current strategy is to wait until you tell me you're done with this PR before looking at anything.

gasche commented:
Sounds good; unfortunately I'm not done with this PR. I hope to get more things done shortly.

gasche committed:
matching: separate types for rows and clauses

"rows" are common abstraction of both pattern analysis and
compilation, but clauses carrying a Lambda.t term are
compilation-specific. Separating rows will thus enable moving more
logic to typing/patterns for use in both settings.

trefis commented:
« shortly » he said :roll_eyes:

gasche commented:
The reason why this is taking longer than I thought it would is that the Simple/Half_simple/General modules in Matching are entangled with matching-only concepts, starting with the notion of "clause" (Parmatch only knows of "rows", and shouldn't depend on the type `lambda`). Disentangling the two parts in a nice way, with small verifiable commits that don't change too much at once, refraining from large API big-bangs at the same time, all this is difficult at I got it wrong several times already.

trefis commented:
:heart:

gasche committed:
reuse ('a Typedtree.pattern_) to define Patterns.Head.t

gasche committed:
matching: move (Non_empty_row, views, General) to patterns.ml

gasche committed:
matching: move {general,simple,half_simple}_view to Patterns

gasche committed:
matching: use pattern views in Parmatch as well

gasche commented:
I felt bad about not giving any visible sign of progress, so I pushed the current state of the branch. As you might notice, more and more type-information changes have been carried into the scope of this PR (I took the decision, I thought it would be an easy change at that point, to have `Head.deconstruct` expect a simple pattern, rather than (unsafely) any pattern). At some point I had to rebase on top of #16 -- so maybe we should indeed think about reviewing that one eventually.

Initially the idea was just "move those definitions to a new module", and at some point along the way it turned into "let's have the best API for this new module ever". A beginner mistake.

I feel like there is still more thread to be pulled, which will inform some changes to the Patterns API -- I think it's time to have a `construct` symmetric to `deconstruct` for example, on top of which the current and now not-flexible-enough `to_omega_pattern` can be expressed.
My natural inclination would be to keep pulling and, when it feels stabilized, look back at what we have and rewrite the history to use that API from the start.

gasche commented:
Ideally the story we would like to tell the upstream repo is:

1. first we create a module Patterns which contains those fine-grained descriptions of intermediate pattern categories and heads, with an API we are proud of
2. then we progressively rewrite Matching and Parmatch to use the features of Patterns, transitioning from program reasoning and sheer luck to type information.

We have done a fairly good share of (2), but are not done yet. I would be happy to stop along the way and send "what we have" upstream for reviewing. But I would like to feel that at least (1) is done properly, and (a) I'm not completely satisfied yet and (b) every time I made a change to improve things, it forced me to do a lot of (2) in reaction.

I know it sounds like a bad time-management approach when I write it down, but I remain tempted to push more on (2) to get (1) to a stable point we like, and then submit (1) upstream, and (2) pieces by pieces.

gasche commented:
(In (2) there is a lot of relatively minor changes, which tend to accumulate a lot of back-and-forth noise as I refine part (1), and will be simplified if we rewrite the story after the fact. But there are also a few big pieces, like many parts of #16, which are relatively big changes and need to be reviewed on their own, I think; you don't want to see them all at once.)

trefis commented:
> A beginner mistake

I think I agree with this part.

For the rest, I understand the process, having been through it myself many times (I just got into such a rabbit hole yesterday!).
However,  I do not quite have the time right now to have a look at what you've done (but I'll find some in the next few weeks). Also, your plan seems like it's going to take a few more months, and might still fail in the end. Unfortunately, I'm afraid I won't have much time for this over the next few months, so I'm very tempted to cut my losses and get out now: I have the feeling that there are already good improvements in what we've written but not yet upstreamed, and I wonder if instead of improving this yet imperfect work, we shouldn't take a more reasonable approach of carving out things that are "acceptable" on their own, right now, and upstream that. Leaving the improvements for ... next year?

gasche commented:
You are the master of branches, I sort of lost count: are there changes not in this PR (which includes `rematch-more-types` and `rematch-simplify-matcher`, with finally `rematch-patterns-in-their-own-file`) that we have not upstreamed yet?

trefis commented:
I... would have to check. I'll look at all of that next week.
